<a href="https://colab.research.google.com/github/mrsferret/Code-Division/blob/main/Worksheets/10_2_Movies_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Movies Mini-project
---

In the previous worksheet you converted an SQL relational database to a single pandas dataframe and downloaded it. You will be analysing it today.

If you were unable to download the file, there is a copy located here: 

"https://github.com/lilaceri/Working-with-data-/blob/main/Data%20Sets%20for%20code%20divisio/movies.csv?raw=true"



### Inspect the dataset 
---

In [145]:
import pandas as pd

def inspect_data(df):

  # firstly, sort the data on movie, director and year
  df = df.sort_values(by = ["name", "last_name", "first_name", "year"])
  display("Describe: ", df.describe())
  display("Head: ", df.head())
  display("Info: ", df.info())

  # check the data for NA/NaN/missing values across the dataframe
  display("Null Values: ", df.isnull().values.any())

  #============================================================================
  # null values do exist in the dataset and df.info() indicates that column "rank"
  # is only column with them
  #============================================================================
  display("Null Values: ", df.isnull().values.any())
  '''' is any missing values in dataframe '''
  display(df.isnull())

  #============================================================================
  # lets look more closely at the rank column as we know it contains null values
  # check what unique values are stored in this column and its count
  #============================================================================
  print("\nUnique values from column rank...\n",df['rank'].unique())
  print("\nCount unique values from a column ...\n",df['rank'].nunique())

  ### get count of missing values of a particular column
 
  #display ("Count of missing values in rank col: ", df.rank.isnull().sum())
  #display (df.rank.isnull().sum())

  display("count of missing values: ", df.isnull().sum())
  display (df.isnull().sum()/len(df)*100)

  #============================================================================
  # lets look more closely at the genre column to check what values are stored 
  # in this column and its count
  #============================================================================

  print("\nUnique values from column genre...\n",df['genre'].unique())
  print("\nCount unique values from a column ...\n",df['genre'].nunique())

  #========================================================================
  # should only be 1 row per movie so check for duplicates
  #========================================================================
  duplicates = find_duplicates(df)
  #duplicates = df[df.duplicated(['name','last_name','year'])]
  #print("Duplicate Values")
  #print(duplicates)
  #print("\nCount unique values from name column ...\n",duplicates['name'].nunique())
  # check if all duplicates have missing values
  #display("count of missing values in duplicates: ", duplicates.isnull().sum())
  #display (duplicates.isnull().sum()/len(duplicates)*100)

  #========================================================================
  # count how many movies have max rating of 9.9
  #========================================================================
  print("Count of movies with 9.9 rank: ",df['rank'].value_counts()[9.9])

def find_duplicates(df):
  #duplicates = df[df.duplicated(['name','last_name','year'])]
  duplicates = df[df.duplicated(['name'],keep=False)]
  display("Duplicate Values")
  display(duplicates.sort_values(by = ["name", "last_name", "year"]))
  display("\nCount unique values from name column ...\n",duplicates['name'].nunique())

  # check if all duplicates have missing values
  display("count of missing values in duplicates: ", duplicates.isnull().sum())
  display (duplicates.isnull().sum()/len(duplicates)*100)

  return duplicates 
  

# program starts here
url = "https://github.com/lilaceri/Working-with-data-/blob/main/Data%20Sets%20for%20code%20divisio/movies.csv?raw=true"

movie_df = pd.read_csv(url)

inspect_data(movie_df)

'Describe: '

Unnamed: 0           year           rank
count  368894.000000  368894.000000  113376.000000
mean   184446.500000    1969.442222       5.995124
std    106490.669438      29.082051       1.540330
min         0.000000    1888.000000       1.000000
25%     92223.250000    1947.000000       5.100000
50%    184446.500000    1976.000000       6.200000
75%    276669.750000    1996.000000       7.100000
max    368893.000000    2007.000000       9.900000

'Head: '

Unnamed: 0 first_name    last_name  \
11330        11330    Mohamed         Asli   
265567      265567      Diane      Pierens   
265568      265568      Diane      Pierens   
265569      265569      Diane      Pierens   
274883      274883  Francisco  Ramalho Jr.   

                                       name  year  rank   genre  
11330    Casablanca les anges ne volent pas  2004   NaN   Drama  
265567                                Clara  1994   NaN   Drama  
265568                                Clara  1994   NaN  Family  
265569                                Clara  1994   NaN   Short  
274883                         Flor da Pele  1976   5.5   Drama

<class 'pandas.core.frame.DataFrame'>
Int64Index: 368894 entries, 11330 to 218388
Data columns (total 7 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   Unnamed: 0  368894 non-null  int64  
 1   first_name  368894 non-null  object 
 2   last_name   368894 non-null  object 
 3   name        368894 non-null  object 
 4   year        368894 non-null  int64  
 5   rank        113376 non-null  float64
 6   genre       368894 non-null  object 
dtypes: float64(1), int64(2), object(4)
memory usage: 22.5+ MB


'Info: '

None

'Null Values: '

True

'Null Values: '

True

Unnamed: 0  first_name  last_name   name   year   rank  genre
11330        False       False      False  False  False   True  False
265567       False       False      False  False  False   True  False
265568       False       False      False  False  False   True  False
265569       False       False      False  False  False   True  False
274883       False       False      False  False  False  False  False
...            ...         ...        ...    ...    ...    ...    ...
170647       False       False      False  False  False   True  False
170648       False       False      False  False  False   True  False
218386       False       False      False  False  False   True  False
218387       False       False      False  False  False   True  False
218388       False       False      False  False  False   True  False

[368894 rows x 7 columns]


Unique values from column rank...
 [nan 5.5 8.7 5.1 5.2 7.8 6.  7.2 4.  6.8 3.1 3.9 6.2 5.8 4.3 6.4 7.3 8.1
 7.5 8.2 3.2 7.9 6.3 5.6 7.1 5.  7.  6.7 6.5 9.6 5.4 6.6 4.9 7.7 6.1 3.5
 6.9 5.7 1.6 4.6 2.  4.8 5.9 3.8 7.4 2.4 4.4 4.7 8.  3.6 5.3 2.8 4.5 2.3
 4.1 8.8 3.3 2.2 9.  7.6 9.8 8.4 9.3 4.2 9.5 8.3 1.4 3.4 2.5 3.7 1.8 9.2
 3.  8.5 2.7 2.9 9.4 1.9 8.9 8.6 9.1 9.7 2.6 1.7 1.  2.1 1.5 1.3 9.9 1.1
 1.2]

Count unique values from a column ...
 90


'count of missing values: '

Unnamed: 0         0
first_name         0
last_name          0
name               0
year               0
rank          255518
genre              0
dtype: int64

Unnamed: 0     0.000000
first_name     0.000000
last_name      0.000000
name           0.000000
year           0.000000
rank          69.265968
genre          0.000000
dtype: float64


Unique values from column genre...
 ['Drama' 'Family' 'Short' 'Comedy' 'Documentary' 'Musical' 'Romance'
 'Fantasy' 'Crime' 'Horror' 'Mystery' 'War' 'Thriller' 'Animation'
 'Western' 'Music' 'Action' 'Sci-Fi' 'Adventure' 'Film-Noir']

Count unique values from a column ...
 20


'Duplicate Values'

Unnamed: 0    first_name last_name              name  year  rank  \
265567      265567         Diane   Pierens             Clara  1994   NaN   
265568      265568         Diane   Pierens             Clara  1994   NaN   
265569      265569         Diane   Pierens             Clara  1994   NaN   
128110      128110  Pierre-Marie    Goulet             Gaule  1974   NaN   
128111      128111  Pierre-Marie    Goulet             Gaule  1974   NaN   
...            ...           ...       ...               ...   ...   ...   
170647      170647         Kemal       Kan  zmc kizin kaderi  1961   NaN   
170648      170648         Kemal       Kan  zmc kizin kaderi  1961   NaN   
218386      218386     Seiji (I)  Maruyama   zora no samurai  1976   NaN   
218387      218387     Seiji (I)  Maruyama   zora no samurai  1976   NaN   
218388      218388     Seiji (I)  Maruyama   zora no samurai  1976   NaN   

              genre  
265567        Drama  
265568       Family  
265569        Short  
128110  Documentary  
128111        Short  
...             ...  
170647        Drama  
170648      Romance  
218386       Action  
218387        Drama  
218388          War  

[270724 rows x 7 columns]

'\nCount unique values from name column ...\n'

99029

'count of missing values in duplicates: '

Unnamed: 0         0
first_name         0
last_name          0
name               0
year               0
rank          181105
genre              0
dtype: int64

Unnamed: 0     0.000000
first_name     0.000000
last_name      0.000000
name           0.000000
year           0.000000
rank          66.896544
genre          0.000000
dtype: float64

Count of movies with 9.9 rank:  55




---


# **Inspecting the Data Notes**

---


1.   On initial look at the data, I can see that the dataset has **368894** rows. From  **.info()**, the only column that contains null data is the "**rank**" column. This will need to be cleaned. The first column in the dataframe is called "**Unnamed: 0**" and on initial inspection it looks like it is just a record number. Will need to look at this further.

2.   **Rank Column** - Looking more closely at the rank column which contains null values, we can see that there are **255518** records with a missing rank value and this represents nearly **70%** of the total number of recs. There are **90** unique values in this column (including nan) with a **max value of 9.9** and a **minimum of 1**. This leaves us with the question of what to do with the rank column? Options are to **a)** drop the column entirely from the dataframe **b)** somehow impute a sensible value for the empty fields **c)** remove only those records that have an empty rank value, although this would result in a dataset **reduced by nearly 70%**, a not insignificant amount.

1.  **Duplicate row check** - when checking for duplicates, no Movie/Director/Year combination should have more than 1 row. When check for rows with duplicate values, **138290** rows are returned (includes all duplicate rows) which represents **89140** rows with a unique movie name. From this initial check, it looks possible that all duplicate rows have missing values in the **rank** column. Check this out.
2.   After checking whether all duplicates have missing values in the **rank** column, I can see that this is not actually the case. Of **138290** duplicate rows returned, only **88465** of these have missing values in the rank field


2.   List item



 






### Clean the dataset 
---


In [122]:
from os import name
import pandas as pd

def clean_data(df):

  # ================================================================
  # remove all rows where column rank is null
  # ================================================================
  clean_df = df[~df["rank"].isnull()]
  #display (clean_df)

  # ================================================================
  # remove leading/trailing spaces from strings
  # ================================================================
 
  # removing whitespace from more than 1 column
  clean_df['name'].str.strip()
  clean_df['first_name'].str.strip()
  clean_df['last_name'].str.strip()

  # ================================================================
  # now that non-nulls removed, remove duplicate movie recs
  # ================================================================
  clean_duplicates = clean_df.drop_duplicates(subset = ['name'], keep = 'last')
  #display ('Cleaned Duplicates: ', clean_duplicates.head(100))

  display("Sorted Duplicate Values: ")
  sorted_df = clean_duplicates.sort_values(by = ["name"])
  #with pd.option_context('display.max_rows', 500,):
  #  print("Sorted: ", sorted_df)
    
  #display("\nCount unique values from name column ...\n",sorted_df['name'].nunique())
  #duplicates = sorted_df[clean_duplicates.duplicated(['name'])]
  #display ("Duplicates after cleaning: ", sorted_df)

  display("Info: ", sorted_df.info())

  return sorted_df
  
# program starts here
url = "https://github.com/lilaceri/Working-with-data-/blob/main/Data%20Sets%20for%20code%20divisio/movies.csv?raw=true"

movie_df = pd.read_csv(url)

clean_data(movie_df)



'Sorted Duplicate Values: '

<class 'pandas.core.frame.DataFrame'>
Int64Index: 55954 entries, 274883 to 151291
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  55954 non-null  int64  
 1   first_name  55954 non-null  object 
 2   last_name   55954 non-null  object 
 3   name        55954 non-null  object 
 4   year        55954 non-null  int64  
 5   rank        55954 non-null  float64
 6   genre       55954 non-null  object 
dtypes: float64(1), int64(2), object(4)
memory usage: 3.4+ MB


'Info: '

None

Unnamed: 0 first_name     last_name  \
274883      274883  Francisco   Ramalho Jr.   
233647      233647     Evaldo      Mocarzel   
233670      233670   Giuseppe        Moccia   
118878      118878     Pierre  Gaspard-Huit   
216153      216153      Mario        Marret   
...            ...        ...           ...   
81940        81940    Jacques          Demy   
1900          1900     Amable       Aguiluz   
20127        20127    Mariano       Barroso   
198580      198580     Gianni         Lepre   
151291      151291     Marius         Holst   

                                                     name  year  rank  \
274883                                       Flor da Pele  1976   5.5   
233647                                   Margem da Imagem  2003   8.7   
233670                              arrivato mio fratello  1985   5.1   
118878                                       belles dents  1966   5.5   
216153                                    bientt, j'espre  1968   5.2   
...                                                   ...   ...   ...   
81940   vnement le plus important depuis que l'homme a...  1973   4.5   
1900                                          www.XXX.com  2003   1.0   
20127                                     xtasis (1996/I)  1996   7.6   
198580                                          ye for ye  1985   4.5   
151291                                         yenstikker  2001   5.3   

              genre  
274883        Drama  
233647  Documentary  
233670       Comedy  
118878        Drama  
216153  Documentary  
...             ...  
81940        Comedy  
1900        Romance  
20127         Drama  
198580        Drama  
151291        Drama  

[55954 rows x 7 columns]

In [95]:
from os import name
import pandas as pd

def misc_data(df):
  return


# program starts here
url = "https://github.com/lilaceri/Working-with-data-/blob/main/Data%20Sets%20for%20code%20divisio/movies.csv?raw=true"

movie_df = pd.read_csv(url)
#df3 = movie_df[(movie_df['last_name'] == 'Aaltonen') & (movie_df['first_name'] == 'Veikko')]
df3 = movie_df
#display (df3)

# ================================================================
# remove all rows where column rank is null
# ================================================================
clean_df = df3[~df3["rank"].isnull()]

# ================================================================
# check for duplicate movies
# ================================================================
clean_df = df3[~df3["rank"].isnull()]
duplicates = clean_df[clean_df.duplicated(['name'],keep=False)]
display ("Duplicates: ", duplicates)
#print(duplicates.to_markdown())
# The scope of these changes
# are local with systems to with statement.
with pd.option_context('display.max_rows', 500,):
    print(duplicates)

display ("Aaltonen Duplicates: ", duplicates[(duplicates['last_name'] == 'Aaltonen')])
display ("Ün Duplicates: ", duplicates[(duplicates['last_name'] == 'Ün')])


misc_data(movie_df)

'Duplicates: '

Unnamed: 0 first_name last_name                    name  year  rank  \
79              79     Veikko  Aaltonen         Juoksuhaudantie  2004   6.3   
80              80     Veikko  Aaltonen         Juoksuhaudantie  2004   6.3   
82              82     Veikko  Aaltonen              Merisairas  1996   4.3   
83              83     Veikko  Aaltonen              Merisairas  1996   4.3   
84              84     Veikko  Aaltonen      Rakkaudella, Maire  1999   6.9   
...            ...        ...       ...                     ...   ...   ...   
368833      368833     Memduh        Ün             Devlet Kusu  1980   7.2   
368837      368837     Memduh        Ün                   Garip  1986   7.9   
368838      368838     Memduh        Ün                   Garip  1986   7.9   
368861      368861     Memduh        Ün  Sezercik aslan parcasi  1972   6.0   
368862      368862     Memduh        Ün  Sezercik aslan parcasi  1972   6.0   

           genre  
79        Comedy  
80         Drama  
82         Drama  
83      Thriller  
84         Drama  
...          ...  
368833     Drama  
368837    Comedy  
368838     Drama  
368861     Drama  
368862   Romance  

[87669 rows x 7 columns]

        Unnamed: 0 first_name last_name                    name  year  rank  \
79              79     Veikko  Aaltonen         Juoksuhaudantie  2004   6.3   
80              80     Veikko  Aaltonen         Juoksuhaudantie  2004   6.3   
82              82     Veikko  Aaltonen              Merisairas  1996   4.3   
83              83     Veikko  Aaltonen              Merisairas  1996   4.3   
84              84     Veikko  Aaltonen      Rakkaudella, Maire  1999   6.9   
...            ...        ...       ...                     ...   ...   ...   
368833      368833     Memduh        Ün             Devlet Kusu  1980   7.2   
368837      368837     Memduh        Ün                   Garip  1986   7.9   
368838      368838     Memduh        Ün                   Garip  1986   7.9   
368861      368861     Memduh        Ün  Sezercik aslan parcasi  1972   6.0   
368862      368862     Memduh        Ün  Sezercik aslan parcasi  1972   6.0   

           genre  
79        Comedy  
80         Dr

'Aaltonen Duplicates: '

Unnamed: 0 first_name last_name                name  year  rank     genre
79          79     Veikko  Aaltonen     Juoksuhaudantie  2004   6.3    Comedy
80          80     Veikko  Aaltonen     Juoksuhaudantie  2004   6.3     Drama
82          82     Veikko  Aaltonen          Merisairas  1996   4.3     Drama
83          83     Veikko  Aaltonen          Merisairas  1996   4.3  Thriller
84          84     Veikko  Aaltonen  Rakkaudella, Maire  1999   6.9     Drama
85          85     Veikko  Aaltonen  Rakkaudella, Maire  1999   6.9  Thriller
86          86     Veikko  Aaltonen           Tilinteko  1987   6.0     Crime
87          87     Veikko  Aaltonen           Tilinteko  1987   6.0     Drama

'Ün Duplicates: '

Unnamed: 0 first_name last_name                    name  year  rank  \
368828      368828     Memduh        Ün                Cevriyem  1978   7.8   
368829      368829     Memduh        Ün                Cevriyem  1978   7.8   
368832      368832     Memduh        Ün             Devlet Kusu  1980   7.2   
368833      368833     Memduh        Ün             Devlet Kusu  1980   7.2   
368837      368837     Memduh        Ün                   Garip  1986   7.9   
368838      368838     Memduh        Ün                   Garip  1986   7.9   
368861      368861     Memduh        Ün  Sezercik aslan parcasi  1972   6.0   
368862      368862     Memduh        Ün  Sezercik aslan parcasi  1972   6.0   

          genre  
368828    Drama  
368829  Romance  
368832   Comedy  
368833    Drama  
368837   Comedy  
368838    Drama  
368861    Drama  
368862  Romance



---


# **Cleaning the Dataset Notes**


---



1.  **Missing Values** - Column **rank** is the only col with missing values. It is a vital piece of info for our data analysis so I'm left with the choice somehow imputing a sensible value into this col or deleting all rows where it has null values or it is non-numeric. As there is no means to impute a sensible value, I have decided that I will need to disregard all rows where rank is not present in all further analysis.

2.   **Cleaning duplicates** - logically there should be 1 record per movie in the dataset, however looking at the data I can see that there are duplicates



### How many movies of each genre are there?
---

In [124]:
import pandas as pd

def movies_by_genre(df):

  #movie_by_genre_cnt = df[["genre"]].groupby("genre").
  movie_by_genre_cnt = df.groupby("genre").count()
  display ("Number of movies by genre: ", movie_by_genre_cnt.name)

  #dept_emp_num =  employee.groupby('DEPT')['DEPT'].count()

  return

# program starts here
url = "https://github.com/lilaceri/Working-with-data-/blob/main/Data%20Sets%20for%20code%20divisio/movies.csv?raw=true"

movie_df = pd.read_csv(url)

df = clean_data(movie_df)
movies_by_genre(df)


  

'Sorted Duplicate Values: '

<class 'pandas.core.frame.DataFrame'>
Int64Index: 55954 entries, 274883 to 151291
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  55954 non-null  int64  
 1   first_name  55954 non-null  object 
 2   last_name   55954 non-null  object 
 3   name        55954 non-null  object 
 4   year        55954 non-null  int64  
 5   rank        55954 non-null  float64
 6   genre       55954 non-null  object 
dtypes: float64(1), int64(2), object(4)
memory usage: 3.4+ MB


'Info: '

None

'Number of movies by genre: '

genre
Action           882
Adventure        898
Animation        175
Comedy          7157
Crime            959
Documentary     2167
Drama          14111
Family          1179
Fantasy          855
Film-Noir        231
Horror          1924
Music            535
Musical         1267
Mystery          979
Romance         4237
Sci-Fi          1765
Short           8143
Thriller        4748
War             1614
Western         2128
Name: name, dtype: int64

### Which director has the highest ranked movies?
---

In [148]:
import pandas as pd

def find_max_rank_director(df):

  #========================================================================
  # count how many movies have max rating of 9.9
  #========================================================================
  print("Count of movies with 9.9 rank: ",df['rank'].value_counts()[9.9])
  display ("Movies with max rating of 9.9: ", df.loc[df['rank'] == 9.9])
  #======================================================================
  # Sort movies by descending rank
  #======================================================================
  sorted_by_rank_df = df.sort_values(by = "rank", ascending=False)

  #======================================================================
  # display movies with max rank of 9.9
  #======================================================================
  N = 60
  display ("Movies sorted by rank 1: ", sorted_by_rank_df.iloc[:N] )

  # 20 largest values in column rank
  sorted_by_rank_df = df.sort_values(by = "rank", ascending=False)
  display ("Movies sorted by rank 2: ", df.nlargest(20, ['rank']) )

  #======================================================================
  # now Sort top 20 movies by director
  #======================================================================
  sorted_by_director_df = sorted_by_rank_df.sort_values(by = ["last_name","first_name","name"], ascending=False)
  display ("Top Movies sorted by Director: ", sorted_by_director_df.iloc[:N] )

  # 20 largest values in column rank
  #df.nlargest(20, ['rank'])
  display ("Movies sorted by rank: ", df.nlargest(20, ['rank']) )
  
  return

# program starts here
url = "https://github.com/lilaceri/Working-with-data-/blob/main/Data%20Sets%20for%20code%20divisio/movies.csv?raw=true"

movie_df = pd.read_csv(url)

df = clean_data(movie_df)
find_max_rank_director(df)

'Sorted Duplicate Values: '

<class 'pandas.core.frame.DataFrame'>
Int64Index: 55954 entries, 274883 to 151291
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  55954 non-null  int64  
 1   first_name  55954 non-null  object 
 2   last_name   55954 non-null  object 
 3   name        55954 non-null  object 
 4   year        55954 non-null  int64  
 5   rank        55954 non-null  float64
 6   genre       55954 non-null  object 
dtypes: float64(1), int64(2), object(4)
memory usage: 3.4+ MB


'Info: '

None

Count of movies with 9.9 rank:  27


'Movies with max rating of 9.9: '

Unnamed: 0  first_name     last_name  \
247643      247643      Sergiu    Nicolaescu   
81826        81826    Jonathan         Demme   
366482      366482       Gregg     Zehentner   
57506        57506    Benjamin   Christensen   
36188        36188   Julio (I)        Bracho   
34473        34473      Satyen          Bose   
289053      289053      Carole  Roussopoulos   
246689      246689    See-Yuen            Ng   
26589        26589     Ishmael        Bernal   
293006      293006      András       Salamon   
94815        94815   Sergei M.    Eisenstein   
163434      163434     Marcell     Jankovics   
356621      356621       Ioana        Wieder   
117235      117235        Abel         Gance   
163728      163728         Jim      Jarmusch   
51113        51113        John    Cassavetes   
67646        67646     Matthew      Crandall   
289096      289096      Carole  Roussopoulos   
72888        72888      Georgi      Daneliya   
337071      337071    Jirí (I)         Trnka   
139694      139694       Fedor    Hanzekovic   
271950      271950    Vsevolod      Pudovkin   
155995      155995  William T.         Hurtz   
366355      366355      Arünas     Zebriünas   
132970      132970      Ivanka     Grybcheva   
47671        47671      Andrew   Cappelletti   
336670      336670     Kirsten       Tretbar   

                                           name  year  rank        genre  
247643  Atunci i-am condamnat pe toti la moarte  1971   9.9          War  
81826                     Complex Sessions, The  1994   9.9        Short  
366482                       Dawn of the Friend  2004   9.9        Short  
57506                       Devil's Circus, The  1926   9.9        Drama  
36188                         Distinto amanecer  1943   9.9     Thriller  
34473                                     Dosti  1964   9.9      Musical  
289053               Genet parle d'Angela Davis  1970   9.9        Short  
246689                           Gong fu qi jie  1979   9.9       Action  
26589                                    Himala  1982   9.9        Drama  
293006                                   Huttyn  1996   9.9  Documentary  
94815                          Ivan Groznyj III  1988   9.9        Drama  
163434                                Jnos vitz  1973   9.9    Animation  
356621           Marche des femmes  Hendaye, La  1975   9.9        Short  
117235                        Napolon Bonaparte  1934   9.9        Drama  
163728                           New World, The  1982   9.9        Short  
51113                          Pair of Boots, A  1962   9.9          War  
67646                          Prince Solitaire  2003   9.9        Short  
289096          Prostitues de Lyon parlent, Les  1975   9.9        Short  
72888                             Slyozy kapali  1982   9.9        Drama  
337071                        Star povesti cesk  1953   9.9    Animation  
139694                      Svoga tela gospodar  1957   9.9        Drama  
271950              Ubijtsi vykhodyat na dorogu  1942   9.9          War  
155995                 Unicorn in the Garden, A  1953   9.9        Short  
366355                           Velnio nuotaka  1973   9.9      Musical  
132970                   Voynata na taralezhite  1979   9.9       Comedy  
47671                            Whirlpool, The  2003   9.9        Short  
336670                                   Zenith  2001   9.9  Documentary

'Movies sorted by rank 1: '

Unnamed: 0     first_name      last_name  \
289053      289053         Carole   Roussopoulos   
57506        57506       Benjamin    Christensen   
81826        81826       Jonathan          Demme   
289096      289096         Carole   Roussopoulos   
271950      271950       Vsevolod       Pudovkin   
293006      293006         András        Salamon   
139694      139694          Fedor     Hanzekovic   
337071      337071       Jirí (I)          Trnka   
132970      132970         Ivanka      Grybcheva   
67646        67646        Matthew       Crandall   
366482      366482          Gregg      Zehentner   
72888        72888         Georgi       Daneliya   
336670      336670        Kirsten        Tretbar   
246689      246689       See-Yuen             Ng   
51113        51113           John     Cassavetes   
163728      163728            Jim       Jarmusch   
356621      356621          Ioana         Wieder   
366355      366355         Arünas      Zebriünas   
94815        94815      Sergei M.     Eisenstein   
117235      117235           Abel          Gance   
34473        34473         Satyen           Bose   
163434      163434        Marcell      Jankovics   
47671        47671         Andrew    Cappelletti   
247643      247643         Sergiu     Nicolaescu   
36188        36188      Julio (I)         Bracho   
26589        26589        Ishmael         Bernal   
155995      155995     William T.          Hurtz   
135109      135109          Rubén          Gámez   
294086      294086  Victoria Rose        Sampson   
133200      133200           Luca   Guardabascio   
170904      170904          Takis  Kanellopoulos   
362500      362500         Eiichi       Yamamoto   
368620      368620        Kentarô          Ôtani   
232228      232228       François          Miron   
270328      270328           L.V.         Prasad   
120217      120217           Jean      Georgescu   
227540      227540     Joao Costa        Menezes   
145917      145917        Jim (I)         Henson   
240190      240190         Mircea        Muresan   
209840      209840     Anthony L.     MacFarland   
30209        30209         Andrei         Blaier   
279364      279364          Jason        Reitman   
309877      309877       Lloyd A.      Silverman   
324615      324615         Yu (I)            Sun   
239233      239233     Hrishikesh      Mukherjee   
98002        98002        Michael         Escher   
116210      116210     Christophe       Gaillard   
240123      240123           Kira       Muratova   
337718      337718    Aristomenis        Tsirbas   
316833      316833         Lionel         Soukaz   
333807      333807  Charles 'Bud'       Tingwell   
316821      316821         Lionel         Soukaz   
271683      271683          Yakov     Protazanov   
289038      289038         Ari M.     Roussimoff   
203374      203374            Wei             Lo   
48559        48559       Patricia        Cardoso   
154904      154904         Oliver        Hummell   
129751      129751      Alfred E.          Green   
154106      154106            Tom       Huckabee   
169805      169805        Djordje      Kadijevic   

                                           name  year  rank        genre  
289053               Genet parle d'Angela Davis  1970   9.9        Short  
57506                       Devil's Circus, The  1926   9.9        Drama  
81826                     Complex Sessions, The  1994   9.9        Short  
289096          Prostitues de Lyon parlent, Les  1975   9.9        Short  
271950              Ubijtsi vykhodyat na dorogu  1942   9.9          War  
293006                                   Huttyn  1996   9.9  Documentary  
139694                      Svoga tela gospodar  1957   9.9        Drama  
337071                        Star povesti cesk  1953   9.9    Animation  
132970                   Voynata na taralezhite  1979   9.9       Comedy  
67646                          Prince Solitaire  2003   9.9        Short  
366482    

'Movies sorted by rank 2: '

Unnamed: 0 first_name     last_name  \
247643      247643     Sergiu    Nicolaescu   
81826        81826   Jonathan         Demme   
366482      366482      Gregg     Zehentner   
57506        57506   Benjamin   Christensen   
36188        36188  Julio (I)        Bracho   
34473        34473     Satyen          Bose   
289053      289053     Carole  Roussopoulos   
246689      246689   See-Yuen            Ng   
26589        26589    Ishmael        Bernal   
293006      293006     András       Salamon   
94815        94815  Sergei M.    Eisenstein   
163434      163434    Marcell     Jankovics   
356621      356621      Ioana        Wieder   
117235      117235       Abel         Gance   
163728      163728        Jim      Jarmusch   
51113        51113       John    Cassavetes   
67646        67646    Matthew      Crandall   
289096      289096     Carole  Roussopoulos   
72888        72888     Georgi      Daneliya   
337071      337071   Jirí (I)         Trnka   

                                           name  year  rank        genre  
247643  Atunci i-am condamnat pe toti la moarte  1971   9.9          War  
81826                     Complex Sessions, The  1994   9.9        Short  
366482                       Dawn of the Friend  2004   9.9        Short  
57506                       Devil's Circus, The  1926   9.9        Drama  
36188                         Distinto amanecer  1943   9.9     Thriller  
34473                                     Dosti  1964   9.9      Musical  
289053               Genet parle d'Angela Davis  1970   9.9        Short  
246689                           Gong fu qi jie  1979   9.9       Action  
26589                                    Himala  1982   9.9        Drama  
293006                                   Huttyn  1996   9.9  Documentary  
94815                          Ivan Groznyj III  1988   9.9        Drama  
163434                                Jnos vitz  1973   9.9    Animation  
356621           Marche des femmes  Hendaye, La  1975   9.9        Short  
117235                        Napolon Bonaparte  1934   9.9        Drama  
163728                           New World, The  1982   9.9        Short  
51113                          Pair of Boots, A  1962   9.9          War  
67646                          Prince Solitaire  2003   9.9        Short  
289096          Prostitues de Lyon parlent, Les  1975   9.9        Short  
72888                             Slyozy kapali  1982   9.9        Drama  
337071                        Star povesti cesk  1953   9.9    Animation

'Top Movies sorted by Director: '

Unnamed: 0  first_name      last_name  \
368891      368891        Idil           Üner   
368862      368862      Memduh             Ün   
368838      368838      Memduh             Ün   
368833      368833      Memduh             Ün   
368829      368829      Memduh             Ün   
368796      368796      Harald   Østgård-Lund   
368791      368791      Anders     Østergaard   
368790      368790      Anders     Østergaard   
368783      368783       Claus         Ørsted   
368779      368779       Peter  Ørskov Madsen   
368769      368769     Henning         Ørnbak   
368767      368767     Henning         Ørnbak   
368758      368758         Pål            Øie   
368753      368753         Pål            Øie   
368751      368751        Ziya          Öztan   
368724      368724       Yavuz          Özkan   
368723      368723       Yavuz          Özkan   
368709      368709         Ali      Özgentürk   
368706      368706         Ali      Özgentürk   
368701      368701       Kazim             Öz   
368694      368694       Peter        Östlund   
368693      368693       Peter        Östlund   
368692      368692       Peter        Östlund   
368671      368671        Zeki          Ökten   
368669      368669        Zeki          Ökten   
368666      368666        Zeki          Ökten   
368665      368665        Zeki          Ökten   
368664      368664        Zeki          Ökten   
368662      368662        Zeki          Ökten   
368661      368661        Zeki          Ökten   
368660      368660        Zeki          Ökten   
368658      368658        Zeki          Ökten   
368657      368657        Zeki          Ökten   
368652      368652        Zeki          Ökten   
368643      368643   Katsuhiro          Ôtomo   
368636      368636   Katsuhiro          Ôtomo   
368631      368631   Katsuhiro          Ôtomo   
368628      368628   Katsuhiro          Ôtomo   
368626      368626   Katsuhiro          Ôtomo   
368620      368620     Kentarô          Ôtani   
368498      368498  Lárus Ýmir      Óskarsson   
368493      368493      Pierre          Étaix   
368492      368492      Pierre          Étaix   
368487      368487      Pierre          Étaix   
368485      368485      Pierre          Étaix   
368432      368432       Marit         Åslein   
368430      368430       Marit         Åslein   
368427      368427       Jonas       Åkerlund   
368426      368426       Jonas       Åkerlund   
368417      368417       Jonas       Åkerlund   
368413      368413         Per          Åhlin   
368409      368409         Per          Åhlin   
368408      368408         Per          Åhlin   
368406      368406         Per          Åhlin   
368400      368400       Lasse          Åberg   
368399      368399       Lasse          Åberg   
368398      368398       Lasse          Åberg   
368397      368397       Lasse          Åberg   
368396      368396       Lasse          Åberg   
368395      368395       Lasse          Åberg   

                                                     name  year  rank  \
368891                 Liebenden vom Hotel von Osman, Die  2001   8.5   
368862                             Sezercik aslan parcasi  1972   6.0   
368838                                              Garip  1986   7.9   
368833                                        Devlet Kusu  1980   7.2   
368829                                           Cevriyem  1978   7.8   
368796                              Halveis til Haugesund  1997   4.7   
368791             Trollkarlen - en film om Jan Johansson  1999   6.5   
368790                                      Tintin et moi  2003   7.6   
368783                                    Prsten i Vejlby  1972   6.1   
368779  Day the Women Took Over or I Was a Teenage Neo...  2001   8.9   
368769                                     Mig og mafiaen  1973   6.5   
368767                         Mafiaen - det er osse mig!  1974   5.2   
368758                                           Villmark  2003   6.9

'Movies sorted by rank: '

Unnamed: 0 first_name     last_name  \
247643      247643     Sergiu    Nicolaescu   
81826        81826   Jonathan         Demme   
366482      366482      Gregg     Zehentner   
57506        57506   Benjamin   Christensen   
36188        36188  Julio (I)        Bracho   
34473        34473     Satyen          Bose   
289053      289053     Carole  Roussopoulos   
246689      246689   See-Yuen            Ng   
26589        26589    Ishmael        Bernal   
293006      293006     András       Salamon   
94815        94815  Sergei M.    Eisenstein   
163434      163434    Marcell     Jankovics   
356621      356621      Ioana        Wieder   
117235      117235       Abel         Gance   
163728      163728        Jim      Jarmusch   
51113        51113       John    Cassavetes   
67646        67646    Matthew      Crandall   
289096      289096     Carole  Roussopoulos   
72888        72888     Georgi      Daneliya   
337071      337071   Jirí (I)         Trnka   

                                           name  year  rank        genre  
247643  Atunci i-am condamnat pe toti la moarte  1971   9.9          War  
81826                     Complex Sessions, The  1994   9.9        Short  
366482                       Dawn of the Friend  2004   9.9        Short  
57506                       Devil's Circus, The  1926   9.9        Drama  
36188                         Distinto amanecer  1943   9.9     Thriller  
34473                                     Dosti  1964   9.9      Musical  
289053               Genet parle d'Angela Davis  1970   9.9        Short  
246689                           Gong fu qi jie  1979   9.9       Action  
26589                                    Himala  1982   9.9        Drama  
293006                                   Huttyn  1996   9.9  Documentary  
94815                          Ivan Groznyj III  1988   9.9        Drama  
163434                                Jnos vitz  1973   9.9    Animation  
356621           Marche des femmes  Hendaye, La  1975   9.9        Short  
117235                        Napolon Bonaparte  1934   9.9        Drama  
163728                           New World, The  1982   9.9        Short  
51113                          Pair of Boots, A  1962   9.9          War  
67646                          Prince Solitaire  2003   9.9        Short  
289096          Prostitues de Lyon parlent, Les  1975   9.9        Short  
72888                             Slyozy kapali  1982   9.9        Drama  
337071                        Star povesti cesk  1953   9.9    Animation

### How many movies have ranks of over 9?
---


### Plot a bar chart of mean rank and genre
---

### Plot a pie chart of how many movies of each genre there are 
---


### Plot a graph showing the mean Rank for each year 

### What else can you find out from this dataset?
---
Make a plan of 3 further things you can do to interrogate and analyse this dataset

Type your answer here  

1.  
2.  
3.   

### Complete the tasks you have set out in the exercise above. 
---


# Reflection
----

## What skills have you demonstrated in completing this notebook?

Your answer:

## What caused you the most difficulty?

Your answer: